src/RequireCallStack.hs

```diff
@@ -48,9 +65,50 @@ import RequireCallStack.Internal
 -- @since 0.1.0.0
 type RequireCallStack = (HasCallStack, RequireCallStackImpl)
 
--- | Raise an 'ErrorCall' and incur a 'RequireCallStack' constraint while
--- you do so. This
+-- | If you're running into this class, then you need to add
+-- 'RequireCallStack' to your function's signature, or discharge the
+-- constraint using 'provideCallStack'.
+--
+-- @since 0.1.0.0
+type RequireCallStackImpl = ?provideCallStack :: ()
+
+-- | Raise an 'Control.Exception.ErrorCall' and incur a 'RequireCallStack'
+-- constraint while you do so. This
 --
 -- @since 0.1.0.0
 errorRequireCallStack :: RequireCallStack => String -> x
 errorRequireCallStack = error
+
+instance TypeError ('Text "Add RequireCallStack to your function context or use provideCallStack") => IP "provideCallStack" ()
```

Comment:
Oh that is very nice!

Reply:
Wait, at what point is this actually invoked?

Reply:
This is invoked in two situations:

- the user defines a (top-level or local) binding whose inferred type would have required `RequireCallStack`

  In other words, if the user does something like:
  
  ```haskell
  main = do
      let myError = errorRequireCallStack "oops!"
      
      fromMaybe myError (Just (print 1))
  ```
  
  … then that would tell the user to add `RequireCallStack` to the context and if the user follows that instruction:
  
  ```haskell
  main = do
      let myError :: RequireCallStack => a
          myError = errorRequireCallStack "oops!"
      
      fromMaybe myError (Just (print 1))
  ```
  
  … then the type error disappears.

- The user defines a (top-level or intermediate) binding that has a type annotation but it is missing a `RequireCallStack`

  In other words, if the user were to do:
  
  ```haskell
  main = do
      let myError :: a
          myError = errorRequireCallStack "oops!"
      
      fromMaybe myError (Just (print 1))
  ```
  
  … then they would get the same error message telling them to add a `RequireCallStack` constraint.